File: 146. LRU 缓存.py
class DLinkedNode:
    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.pre = None
        self.next = None


class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = {}

        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.pre = self.head

    def _remove_node(self, node):
        """物理切断：把某个节点从链表中抠出来"""
        node.next.pre = node.pre
        node.pre.next = node.next

    def _add_to_head(self, node):
        """VIP插队: 把节点塞到保安 Head 的正后方（最安全的位置）"""
        node.next = self.head.next
        node.pre = self.head
        self.head.next.pre = node
        self.head.next = node

    def _move_to_head(self, node):
        """组合技：抠出来，再插到最前面"""
        self._remove_node(node)
        self._add_to_head(node)

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._move_to_head(node)
        return node.value

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._move_to_head(node)
        else:
            new_node = DLinkedNode(key, value)
            self.cache[key] = new_node
            self._add_to_head(new_node)
            if len(self.cache) > self.capacity:
                loser = self.tail.pre
                self._remove_node(loser)
                del self.cache[loser.key]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146. LRU 缓存.go
package main

type Node struct {
	Key   int
	Value int
	Pre   *Node
	Next  *Node
}

type LRUCache struct {
	Capacity int
	Cache    map[int]*Node
	Head     *Node
	Tail     *Node
}

func Constructor(capacity int) LRUCache {
	head := &Node{}
	tail := &Node{}
	head.Next, tail.Pre = tail, head

	return LRUCache{
		Capacity: capacity,
		Cache:    make(map[int]*Node),
		Head:     head,
		Tail:     tail,
	}
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.Cache[key]; ok {
		this.moveToHead(node)
		return node.Value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.Cache[key]; ok {
		node.Value = value
		this.moveToHead(node)
		return
	}

	if len(this.Cache) == this.Capacity {
		lastKey := this.deleteLast()
		delete(this.Cache, lastKey)
	}

	newNode := &Node{
		Key:   key,
		Value: value,
	}
	this.addToHead(newNode)
	this.Cache[key] = newNode
}

func (this *LRUCache) moveToHead(node *Node) {
	node.Next.Pre, node.Pre.Next = node.Pre, node.Next
	node.Next = this.Head.Next
	node.Next.Pre = node
	this.Head.Next = node
	node.Pre = this.Head
}

func (this *LRUCache) deleteLast() int {
	last := this.Tail.Pre
	this.Tail.Pre.Pre.Next = this.Tail
	this.Tail.Pre = this.Tail.Pre.Pre
	return last.Key
}

func (this *LRUCache) addToHead(node *Node) {
	node.Next = this.Head.Next
	node.Pre = this.Head
	this.Head.Next.Pre = node
	this.Head.Next = node
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
